README.rst
=====
aioh2
=====

.. image:: https://img.shields.io/pypi/v/aioh2.svg
        :target: https://pypi.python.org/pypi/aioh2

.. image:: https://img.shields.io/travis/decentfox/aioh2.svg
        :target: https://travis-ci.org/decentfox/aioh2

.. image:: https://readthedocs.org/projects/aioh2/badge/?version=latest
        :target: https://readthedocs.org/projects/aioh2/?badge=latest
        :alt: Documentation Status


HTTP/2 implementation with hyper-h2_ on Python 3 asyncio.

* Free software: BSD license
* Documentation: https://aioh2.readthedocs.org.

Features
--------

* Asynchronous HTTP/2 client and server
* Multiplexing streams of data with managed flow and priority control
* Optional tickless health check
* More to come

Non-features:

* Request/Response wrappers
* Web server, dispatcher, cookie, etc
* HTTP/2 upgrade

Example
-------

A server saying hello:

.. code-block:: python

    # Server request handler
    async def on_connected(proto):
        try:
            while True:
                # Receive a request from queue
                stream_id, headers = await proto.recv_request()

                # Send response headers
                await proto.start_response(stream_id, {':status': '200'})

                # Response body starts with "hello"
                await proto.send_data(stream_id, b'hello, ')

                # Read all request body as a name from client side
                name = await proto.read_stream(stream_id, -1)

                # Amend response body with the name
                await proto.send_data(stream_id, name)

                # Send trailers with the length of the name
                await proto.send_trailers(stream_id, {'len': str(len(name))})
        except asyncio.CancelledError:
            # do cleanup here
            pass

    # Start server on random port, with maximum concurrent requests of 3
    server = await aioh2.start_server(on_connected, port=0, concurrency=3)
    port = server.sockets[0].getsockname()[1]


And a client to try out:

.. code-block:: python

    # Open client connection
    client = await aioh2.open_connection('0.0.0.0', port,
                                         functional_timeout=0.1)

    # Optionally wait for an ack of tickless ping - a.k.a. until functional
    await asyncio.sleep(0.1)  # simulate client being busy with something else
    rtt = await client.wait_functional()
    if rtt:
        print('Round-trip time: %.1fms' % (rtt * 1000))

    # Start request with headers
    stream_id = await client.start_request(
        {':method': 'GET', ':path': '/index.html'})

    # Send my name "world" as whole request body
    await client.send_data(stream_id, b'world', end_stream=True)

    # Receive response headers
    headers = await client.recv_response(stream_id)
    print('Response headers:', headers)

    # Read all response body
    resp = await client.read_stream(stream_id, -1)
    print('Response body:', resp)

    # Read response trailers
    trailers = await client.recv_trailers(stream_id)
    print('Response trailers:', trailers)

    client.close_connection()
    await asyncio.sleep(.1)


Above example can be found at `examples/core.py`.


Credits
-------

A big thanks to the great library hyper-h2_ from `Cory Benfield`_.

`DecentFoX Studio`_ is a software outsourcing company delivering high-quality
web-based products and mobile apps for global customers with agile methodology,
focusing on bleeding-edge technologies and fast-developing scalable architectures.

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage
.. _hyper-h2: https://github.com/python-hyper/hyper-h2
.. _`DecentFoX Studio`: http://decentfox.com
.. _`Cory Benfield`: https://github.com/Lukasa


HISTORY.rst
=======
History
=======

0.2.1 (2017-12-03)
------------------

* Updated to h2 3.0 (Contributed by Arthur Darcet and Peter Wu)
* Dropped support for Python 3.3

0.1.0 (2016-2-6)
----------------

* First release on PyPI.


tests/__init__.py
import functools
import os
import unittest
import uuid

import asyncio
from h2.config import H2Configuration
from h2.connection import H2Connection
from h2.events import RemoteSettingsChanged, RequestReceived, WindowUpdated
from h2.events import SettingsAcknowledged

from aioh2 import H2Protocol


def async_test(timeout=1):
    func = None
    if callable(timeout):
        func = timeout
        timeout = 1

    def _decorator(f):
        @functools.wraps(f)
        def _wrapper(self, *args, **kwargs):
            task = self.loop.create_task(
                asyncio.coroutine(f)(self, *args, **kwargs))

            def _cancel():
                task.print_stack()
                task.cancel()

            time_handle = self.loop.call_later(timeout, _cancel)
            try:
                return self.loop.run_until_complete(task)
            except asyncio.CancelledError:
                events = []
                while True:
                    try:
                        events.append(self.server.events.get_nowait())
                    except asyncio.QueueEmpty:
                        break
                self.fail('server events: {}'.format(events))
            finally:
                time_handle.cancel()

        return _wrapper

    if func is not None:
        return _decorator(func)

    return _decorator


class Server(H2Protocol):
    def __init__(self, test, client_side, *, loop=None):
        super().__init__(client_side, loop=loop)
        test.server = self
        self.events = asyncio.Queue()

    def _event_received(self, event):
        self.events.put_nowait(event)
        super()._event_received(event)


class BaseTestCase(unittest.TestCase):
    def setUp(self):
        self.loop = asyncio.get_event_loop()
        self.path = os.path.join('/tmp', uuid.uuid4().hex)
        self.server = None
        self._server = self.loop.run_until_complete(
            self.loop.create_unix_server(
                lambda: Server(self, False, loop=self.loop), self.path))
        self.loop.run_until_complete(self._setUp())

    def tearDown(self):
        self._server.close()
        os.remove(self.path)
        self.w.close()

    @asyncio.coroutine
    def _setUp(self):
        self.r, self.w = yield from asyncio.open_unix_connection(self.path)
        config = H2Configuration(header_encoding='utf-8')
        self.conn = H2Connection(config=config)
        self.conn.initiate_connection()
        self.w.write(self.conn.data_to_send())
        events = yield from self._expect_events(3)
        self.assertIsInstance(events[0], RemoteSettingsChanged)
        self.assertIsInstance(events[1], RemoteSettingsChanged)
        self.assertIsInstance(events[2], SettingsAcknowledged)

        self.assertIsInstance((yield from self.server.events.get()),
                              RemoteSettingsChanged)
        self.assertIsInstance((yield from self.server.events.get()),
                              SettingsAcknowledged)
        self.assertIsInstance((yield from self.server.events.get()),
                              SettingsAcknowledged)

    @asyncio.coroutine
    def _expect_events(self, n=1):
        events = []
        self.w.write(self.conn.data_to_send())
        while len(events) < n:
            events += self.conn.receive_data((yield from self.r.read(1024)))
            self.w.write(self.conn.data_to_send())
        self.assertEqual(len(events), n)
        return events

    @asyncio.coroutine
    def _send_headers(self, end_stream=False):
        headers = [
            (':method', 'GET'),
            (':authority', 'example.com'),
            (':scheme', 'h2c'),
            (':path', '/index.html'),
        ]
        stream_id = self.conn.get_next_available_stream_id()
        self.conn.send_headers(stream_id, headers, end_stream=end_stream)
        yield from self._expect_events(0)
        event = yield from self.server.events.get()
        self.assertIsInstance(event, RequestReceived)
        self.assertEqual(event.stream_id, stream_id)
        self.assertEqual(event.headers, headers)
        return stream_id

    @asyncio.coroutine
    def _expect_connection_flow_control_disabled(self):
        events = yield from self._expect_events()
        self.assertIsInstance(events[0], WindowUpdated)
        self.assertEqual(events[0].stream_id, 0)
        self.assertGreater(events[0].delta, 1073741822)

    @asyncio.coroutine
    def _assert_received(self, stream_id, coro, expected):
        data = yield from coro
        self.assertEqual(data, expected)

        ack = 0
        while ack < len(data):
            events = yield from self._expect_events()
            self.assertIsInstance(events[0], WindowUpdated)
            self.assertEqual(events[0].stream_id, stream_id)
            ack += events[0].delta
        self.assertEqual(ack, len(data))


tests/test_aioh2.py
#!/usr/bin/env python

"""
test_aioh2
----------------------------------

Tests for `aioh2` module.
"""
import random
import unittest
import uuid

import asyncio
from h2.events import DataReceived
from h2.events import PingAcknowledged
from h2.events import RemoteSettingsChanged
from h2.events import ResponseReceived
from h2.events import SettingsAcknowledged
from h2.exceptions import FlowControlError
from h2.settings import SettingCodes

from aioh2 import SendException
from aioh2.helper import async_task
from . import async_test, BaseTestCase


class TestServer(BaseTestCase):
    def test_connect(self):
        pass

    @async_test
    def test_ping(self):
        opaque_data = uuid.uuid4().bytes[:8]
        self.conn.ping(opaque_data)
        events = yield from self._expect_events()
        self.assertIsInstance(events[0], PingAcknowledged)
        self.assertEqual(events[0].ping_data, opaque_data)

    @async_test
    def test_request_headers(self):
        yield from self._send_headers()

    @asyncio.coroutine
    def _test_read_frame(self, *, more, end_stream):
        stream_id = yield from self._send_headers()

        data = b'x' * random.randint(128, 512)
        self.conn.send_data(stream_id, data,
                            end_stream=not more and end_stream)
        extra = b''
        if more:
            extra = b'y' * random.randint(128, 512)
            self.conn.send_data(stream_id, extra, end_stream=end_stream)

        yield from self._expect_connection_flow_control_disabled()

        yield from self._assert_received(
            stream_id, self.server.read_stream(stream_id), data)

        if more:
            yield from self._assert_received(
                stream_id, self.server.read_stream(stream_id), extra)

        if end_stream:
            frame = yield from self.server.read_stream(stream_id)
            self.assertEqual(frame, b'')
        else:
            try:
                yield from asyncio.wait_for(
                    self.server.read_stream(stream_id), 0.1)
            except asyncio.TimeoutError:
                pass
            else:
                self.assertRaises(asyncio.TimeoutError, lambda: None)

    @async_test
    def test_read_frame(self):
        yield from self._test_read_frame(more=True, end_stream=False)

    @async_test
    def test_read_frame_close(self):
        yield from self._test_read_frame(more=True, end_stream=True)

    @async_test
    def test_read_only_frame(self):
        yield from self._test_read_frame(more=False, end_stream=False)

    @async_test
    def test_read_only_frame_close(self):
        yield from self._test_read_frame(more=False, end_stream=True)

    @asyncio.coroutine
    def _test_read_all(self, *, more, end_stream):
        stream_id = yield from self._send_headers()

        data = b'x' * random.randint(128, 512)
        self.conn.send_data(stream_id, data,
                            end_stream=not more and end_stream)
        if more:
            extra = b'y' * random.randint(128, 512)
            self.conn.send_data(stream_id, extra, end_stream=end_stream)
            data += extra

        yield from self._expect_connection_flow_control_disabled()

        if end_stream:
            yield from self._assert_received(
                stream_id, self.server.read_stream(stream_id, -1), data)

            frame = yield from self.server.read_stream(stream_id, -1)
            self.assertEqual(frame, b'')
        else:
            try:
                yield from asyncio.wait_for(
                    self.server.read_stream(stream_id, -1), 0.1)
            except asyncio.TimeoutError:
                pass
            else:
                self.assertRaises(asyncio.TimeoutError, lambda: None)

    @async_test
    def test_read_all(self):
        yield from self._test_read_all(more=True, end_stream=False)

    @async_test
    def test_read_all_close(self):
        yield from self._test_read_all(more=True, end_stream=True)

    @async_test
    def test_read_all_only_frame(self):
        yield from self._test_read_all(more=False, end_stream=False)

    @async_test
    def test_read_all_only_frame_close(self):
        yield from self._test_read_all(more=False, end_stream=True)

    @asyncio.coroutine
    def _test_read_exactly(self, *, empty, explicit_close):
        stream_id = yield from self._send_headers()

        self.conn.send_data(stream_id, b'333')
        self.conn.send_data(stream_id, b'55555')
        if empty:
            self.conn.send_data(stream_id, b'')

        yield from self._expect_connection_flow_control_disabled()

        yield from self._assert_received(
            stream_id, self.server.read_stream(stream_id, 2), b'33')

        self.conn.send_data(stream_id, b'88888888',
                            end_stream=not explicit_close)
        if explicit_close:
            self.conn.end_stream(stream_id)
        yield from self._expect_events(0)

        yield from self._assert_received(
            stream_id, self.server.read_stream(stream_id, 8), b'35555588')
        yield from self._assert_received(
            stream_id, self.server.read_stream(stream_id, 2), b'88')
        yield from self._assert_received(
            stream_id, self.server.read_stream(stream_id, 8), b'8888')

    @async_test
    def test_read_exactly(self):
        yield from self._test_read_exactly(empty=False, explicit_close=False)

    @async_test
    def test_read_exactly_empty_frame(self):
        yield from self._test_read_exactly(empty=True, explicit_close=False)

    @async_test
    def test_read_exactly_explicit_close(self):
        yield from self._test_read_exactly(empty=True, explicit_close=True)

    @async_test
    def test_read_exactly_empty_frame_explicit_close(self):
        yield from self._test_read_exactly(empty=True, explicit_close=True)

    @async_test
    def test_flow_control_settings(self):
        self.server.update_settings({SettingCodes.INITIAL_WINDOW_SIZE: 3})
        event = yield from self._expect_events()
        self.assertIsInstance(event[0], RemoteSettingsChanged)
        event = yield from self.server.events.get()
        self.assertIsInstance(event, SettingsAcknowledged)

        stream_id = yield from self._send_headers()
        self.conn.send_data(stream_id, b'xx')

        yield from self._expect_connection_flow_control_disabled()

        yield from self._assert_received(
            stream_id, self.server.read_stream(stream_id, 2), b'xx')

        self.conn.send_data(stream_id, b'xxx')
        yield from self._expect_events(0)
        yield from self._assert_received(
            stream_id, self.server.read_stream(stream_id, 3), b'xxx')

        self.assertRaises(FlowControlError,
                          self.conn.send_data, stream_id, b'xxxx')

    @async_test
    def test_flow_control(self):
        self.conn.update_settings({SettingCodes.INITIAL_WINDOW_SIZE: 3})
        event = yield from self._expect_events()
        self.assertIsInstance(event[0], SettingsAcknowledged)
        event = yield from self.server.events.get()
        self.assertIsInstance(event, RemoteSettingsChanged)

        stream_id = yield from self._send_headers(end_stream=True)
        yield from self.server.start_response(stream_id, [(':status', '200')])
        events = yield from self._expect_events()
        self.assertIsInstance(events[0], ResponseReceived)

        yield from self.server.send_data(stream_id, b'12')
        events = yield from self._expect_events()
        self.assertIsInstance(events[0], DataReceived)
        self.assertEqual(events[0].data, b'12')

        try:
            yield from asyncio.wait_for(
                self.server.send_data(stream_id, b'34'), 0.1)
        except asyncio.TimeoutError:
            events = yield from self._expect_events()
            self.assertIsInstance(events[0], DataReceived)
            self.assertEqual(events[0].data, b'3')
        else:
            self.assertRaises(asyncio.TimeoutError, lambda: None)

        self.conn.increment_flow_control_window(3, stream_id=stream_id)
        yield from self._expect_events(0)

        try:
            yield from asyncio.wait_for(
                self.server.send_data(stream_id, b'5678'), 0.1)
        except asyncio.TimeoutError:
            events = yield from self._expect_events()
            self.assertIsInstance(events[0], DataReceived)
            self.assertEqual(events[0].data, b'567')
        else:
            self.assertRaises(asyncio.TimeoutError, lambda: None)

    @async_test
    def test_broken_send(self):
        self.conn.update_settings({SettingCodes.INITIAL_WINDOW_SIZE: 3})
        event = yield from self._expect_events()
        self.assertIsInstance(event[0], SettingsAcknowledged)
        event = yield from self.server.events.get()
        self.assertIsInstance(event, RemoteSettingsChanged)

        stream_id = yield from self._send_headers(end_stream=True)
        yield from self.server.start_response(stream_id, [(':status', '200')])
        events = yield from self._expect_events()
        self.assertIsInstance(events[0], ResponseReceived)

        yield from self.server.send_data(stream_id, b'12')
        events = yield from self._expect_events()
        self.assertIsInstance(events[0], DataReceived)
        self.assertEqual(events[0].data, b'12')

        f = async_task(self.server.send_data(stream_id, b'345678'))
        events = yield from self._expect_events()
        self.assertIsInstance(events[0], DataReceived)
        self.assertEqual(events[0].data, b'3')

        self.conn.reset_stream(stream_id)
        yield from self._expect_events(0)

        try:
            yield from f
        except SendException as e:
            self.assertEqual(e.data, b'45678')
        else:
            self.assertRaises(SendException, lambda: None)

    @async_test(timeout=8)
    def test_priority(self):
        self.conn.update_settings({
            SettingCodes.MAX_FRAME_SIZE: 16384,
            SettingCodes.INITIAL_WINDOW_SIZE: 16384 * 1024 * 32,
        })
        event = yield from self._expect_events()
        self.assertIsInstance(event[0], SettingsAcknowledged)
        event = yield from self.server.events.get()
        self.assertIsInstance(event, RemoteSettingsChanged)

        stream_1 = yield from self._send_headers()
        yield from self.server.start_response(stream_1, [(':status', '200')])
        events = yield from self._expect_events()
        self.assertIsInstance(events[0], ResponseReceived)

        stream_2 = yield from self._send_headers()
        yield from self.server.start_response(stream_2, [(':status', '200')])
        events = yield from self._expect_events()
        self.assertIsInstance(events[0], ResponseReceived)

        p1 = 32
        p2 = 20

        self.server.reprioritize(stream_1, weight=p1)
        self.server.reprioritize(stream_2, weight=p2)
        self.server.pause_writing()

        running = [True]

        @asyncio.coroutine
        def _write(stream_id):
            count = 0
            while running[0]:
                yield from self.server.send_data(stream_id, b'x')
                count += 1
            yield from self.server.end_stream(stream_id)
            return count

        task_1 = async_task(_write(stream_1))
        task_2 = async_task(_write(stream_2))

        for i in range(1000):
            self.server.resume_writing()
            yield from asyncio.sleep(0.004)
            self.server.pause_writing()
            yield from asyncio.sleep(0.001)

        running[0] = False
        self.server.resume_writing()

        count_1 = yield from task_1
        count_2 = yield from task_2

        self.assertAlmostEqual(count_1 / count_2, p1 / p2, 1)


if __name__ == '__main__':
    import sys

    sys.exit(unittest.main())


tests/test_helper.py
import uuid

import asyncio

import aioh2
from . import BaseTestCase, async_test


class TestHelper(BaseTestCase):
    @asyncio.coroutine
    def _cb(self, proto):
        while True:
            stream_id, headers = yield from proto.recv_request()
            # print('  < REQ {}'.format(stream_id))
            yield from proto.start_response(stream_id, [(':status', '200')])
            yield from proto.send_data(stream_id, b'hello, ')
            resp = yield from proto.read_stream(stream_id, -1)
            yield from proto.send_data(stream_id, resp)
            # yield from asyncio.sleep(0.1)
            # print('  > REP {}'.format(stream_id))
            yield from proto.send_trailers(stream_id, [('len', str(len(resp)))])

    @async_test
    def test_tcp(self):
        # Start server
        # FIXME concurrency=3 should work and block the consumer. Figure out why
        # it does not work.
        server = yield from aioh2.start_server(self._cb, host='127.0.0.1',
                                               port=0, concurrency=8)
        port = server.sockets[0].getsockname()[1]

        client = yield from aioh2.open_connection('127.0.0.1', port)
        client.functional_timeout = 0.1
        yield from asyncio.sleep(0.2)

        @asyncio.coroutine
        def _test():
            rtt = yield from client.wait_functional()
            self.assertIsNotNone(rtt)
            self.assertNotAlmostEqual(rtt, self.loop.time(), places=1)
            # Send request
            stream_id = yield from client.start_request([
                (':scheme', 'h2c'),
                (':authority', 'example.com'),
                (':method', 'GET'),
                (':path', '/index.html'),
            ])
            # print('> REQ {}'.format(stream_id))
            name = uuid.uuid4().bytes
            yield from client.send_data(stream_id, name, end_stream=True)

            # Receive response
            headers = yield from client.recv_response(stream_id)
            self.assertEqual(dict(headers)[':status'], '200')
            resp = yield from client.read_stream(stream_id, -1)
            # print('< REP {}'.format(stream_id))
            self.assertEqual(resp, b'hello, ' + name)
            trailers = yield from client.recv_trailers(stream_id)
            self.assertEqual(dict(trailers)['len'], str(len(name)))

        fs, _ = yield from asyncio.wait([_test() for _ in range(8)])
        for f in fs:
            f.result()
        client.close_connection()
        yield from asyncio.sleep(0.1)


aioh2/helper.py
import socket

import asyncio

from .protocol import H2Protocol

__all__ = ['open_connection', 'start_server']
if hasattr(socket, 'AF_UNIX'):
    __all__.extend(['open_unix_connection', 'start_unix_server'])

PROTOCOL_KWARGS = ('concurrency', 'functional_timeout')


class _None:
    pass


def _split_kwargs(kwargs):
    rv = {}
    for key in PROTOCOL_KWARGS:
        val = kwargs.pop(key, _None)
        if val is not _None:
            rv[key] = val
    return rv


@asyncio.coroutine
def open_connection(host=None, port=None, *, loop=None, **kwargs):
    if loop is None:
        loop = asyncio.get_event_loop()
    # noinspection PyArgumentList
    rv = H2Protocol(True, loop=loop, **_split_kwargs(kwargs))
    # noinspection PyArgumentList
    yield from loop.create_connection(lambda: rv, host, port, **kwargs)
    return rv


@asyncio.coroutine
def start_server(client_connected_cb, host=None, port=None, *, loop=None,
                 **kwargs):
    if loop is None:
        loop = asyncio.get_event_loop()

    args = _split_kwargs(kwargs)

    def factory():
        # noinspection PyArgumentList
        rv = H2Protocol(False, loop=loop, **args)
        rv.set_handler(client_connected_cb(rv))
        return rv

    # noinspection PyArgumentList
    return (yield from loop.create_server(factory, host, port, **kwargs))


if hasattr(socket, 'AF_UNIX'):
    @asyncio.coroutine
    def open_unix_connection(path=None, *, loop=None, **kwargs):
        if loop is None:
            loop = asyncio.get_event_loop()
        # noinspection PyArgumentList
        rv = H2Protocol(True, loop=loop, **_split_kwargs(kwargs))
        # noinspection PyArgumentList
        yield from loop.create_unix_connection(lambda: rv, path, **kwargs)
        return rv

    @asyncio.coroutine
    def start_unix_server(client_connected_cb, path=None, *, loop=None,
                          **kwargs):
        if loop is None:
            loop = asyncio.get_event_loop()

        args = _split_kwargs(kwargs)

        def factory():
            # noinspection PyArgumentList
            rv = H2Protocol(False, loop=loop, **args)
            rv.set_handler(client_connected_cb(rv))
            return rv

        # noinspection PyArgumentList
        return (yield from loop.create_unix_server(factory, path, **kwargs))


if hasattr(asyncio, 'ensure_future'):  # Python >= 3.5
    async_task = asyncio.ensure_future
else:
    async_task = asyncio.async


travis_pypi_setup.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Update encrypted deploy password in Travis config file
"""


from __future__ import print_function
import base64
import json
import os
from getpass import getpass
import yaml
from cryptography.hazmat.primitives.serialization import load_pem_public_key
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric.padding import PKCS1v15


try:
    from urllib import urlopen
except:
    from urllib.request import urlopen


GITHUB_REPO = 'decentfox/aioh2'
TRAVIS_CONFIG_FILE = os.path.join(
    os.path.dirname(os.path.abspath(__file__)), '.travis.yml')


def load_key(pubkey):
    """Load public RSA key, with work-around for keys using
    incorrect header/footer format.

    Read more about RSA encryption with cryptography:
    https://cryptography.io/latest/hazmat/primitives/asymmetric/rsa/
    """
    try:
        return load_pem_public_key(pubkey.encode(), default_backend())
    except ValueError:
        # workaround for https://github.com/travis-ci/travis-api/issues/196
        pubkey = pubkey.replace('BEGIN RSA', 'BEGIN').replace('END RSA', 'END')
        return load_pem_public_key(pubkey.encode(), default_backend())


def encrypt(pubkey, password):
    """Encrypt password using given RSA public key and encode it with base64.

    The encrypted password can only be decrypted by someone with the
    private key (in this case, only Travis).
    """
    key = load_key(pubkey)
    encrypted_password = key.encrypt(password, PKCS1v15())
    return base64.b64encode(encrypted_password)


def fetch_public_key(repo):
    """Download RSA public key Travis will use for this repo.

    Travis API docs: http://docs.travis-ci.com/api/#repository-keys
    """
    keyurl = 'https://api.travis-ci.org/repos/{0}/key'.format(repo)
    data = json.loads(urlopen(keyurl).read().decode())
    if 'key' not in data:
        errmsg = "Could not find public key for repo: {}.\n".format(repo)
        errmsg += "Have you already added your GitHub repo to Travis?"
        raise ValueError(errmsg)
    return data['key']


def prepend_line(filepath, line):
    """Rewrite a file adding a line to its beginning.
    """
    with open(filepath) as f:
        lines = f.readlines()

    lines.insert(0, line)

    with open(filepath, 'w') as f:
        f.writelines(lines)


def load_yaml_config(filepath):
    with open(filepath) as f:
        return yaml.load(f)


def save_yaml_config(filepath, config):
    with open(filepath, 'w') as f:
        yaml.dump(config, f, default_flow_style=False)


def update_travis_deploy_password(encrypted_password):
    """Update the deploy section of the .travis.yml file
    to use the given encrypted password.
    """
    config = load_yaml_config(TRAVIS_CONFIG_FILE)

    config['deploy']['password'] = dict(secure=encrypted_password)

    save_yaml_config(TRAVIS_CONFIG_FILE, config)

    line = ('# This file was autogenerated and will overwrite'
            ' each time you run travis_pypi_setup.py\n')
    prepend_line(TRAVIS_CONFIG_FILE, line)


def main(args):
    public_key = fetch_public_key(args.repo)
    password = args.password or getpass('PyPI password: ')
    update_travis_deploy_password(encrypt(public_key, password.encode()))
    print("Wrote encrypted password to .travis.yml -- you're ready to deploy")


if '__main__' == __name__:
    import argparse
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--repo', default=GITHUB_REPO,
                        help='GitHub repo (default: %s)' % GITHUB_REPO)
    parser.add_argument('--password',
                        help='PyPI password (will prompt if not provided)')

    args = parser.parse_args()
    main(args)


examples/core.py
import asyncio
import aioh2

async def example():
    # Server request handler
    async def on_connected(proto):
        try:
            while True:
                # Receive a request from queue
                stream_id, headers = await proto.recv_request()

                # Send response headers
                await proto.start_response(stream_id, [(':status', '200')])

                # Response body starts with "hello"
                await proto.send_data(stream_id, b'hello, ')

                # Read all request body as a name from client side
                name = await proto.read_stream(stream_id, -1)

                # Amend response body with the name
                await proto.send_data(stream_id, name)

                # Send trailers with the length of the name
                await proto.send_trailers(stream_id, [('len', str(len(name)))])
        except asyncio.CancelledError:
            print('Task terminated, saving to disk...')

    # Start server on random port, with maximum concurrent requests of 3
    server = await aioh2.start_server(on_connected, host='127.0.0.1', port=0,
                                      concurrency=3)
    port = server.sockets[0].getsockname()[1]

    # Open client connection
    client = await aioh2.open_connection('127.0.0.1', port,
                                         functional_timeout=0.1)

    # Optionally wait for an ack of tickless ping - a.k.a. until functional
    await asyncio.sleep(0.1)  # simulate a busy client with something else
    rtt = await client.wait_functional()
    if rtt:
        print('Round-trip time: %.1fms' % (rtt * 1000))

    # Start request with headers
    stream_id = await client.start_request([
        (':scheme', 'h2c'),
        (':authority', 'example.com'),
        (':method', 'GET'),
        (':path', '/index.html'),
    ])

    # Send my name "world" as whole request body
    await client.send_data(stream_id, b'world', end_stream=True)

    # Receive response headers
    headers = await client.recv_response(stream_id)
    print('Response headers:', headers)

    # Read all response body
    resp = await client.read_stream(stream_id, -1)
    print('Response body:', resp)

    # Read response trailers
    trailers = await client.recv_trailers(stream_id)
    print('Response trailers:', trailers)

    client.close_connection()
    await asyncio.sleep(.1)


asyncio.get_event_loop().run_until_complete(example())
